pallets/validators-manager/src/lib.rs

```diff
@@ -640,26 +565,237 @@ impl<T: Config> Pallet<T> {
+        let mut found_entry: Option<(T::AccountId, IngressCounter, ValidatorsActionType)> = None;
+        
+        for (account_id, ingress_counter, validators_action_data) in <ValidatorActions<T>>::iter() {
+            if validators_action_data.eth_transaction_id == tx_id {
+                found_entry = Some((
+                    account_id,
+                    ingress_counter,
+                    validators_action_data.action_type,
+                ));
+                break;
+            }
+        }
+
+        let Some((account_id, ingress_counter, action_type)) = found_entry else {
+            // No matching entry found - might have been cleaned up already
+            return Ok(())
+        };
+
+        if !succeeded {
+            // T1 operation failed - cleanup the action and rollback state
+            if action_type.is_registration() {
+                // Remove the eth key mapping we added optimistically
+                if let Some(eth_key) = Self::get_ethereum_public_key_if_exists(&account_id) {
+                    <EthereumPublicKeys<T>>::remove(eth_key);
+                }
+                // Cleanup stored deposit
+                PendingRegistrationDeposits::<T>::remove(&account_id);
             }
+            
+            <ValidatorActions<T>>::remove(&account_id, ingress_counter);
+            Self::deposit_event(Event::<T>::PublishingValidatorActionOnEthereumFailed { tx_id });
+            return Ok(())
+        }
+
+        // T1 succeeded - complete the operation based on action type
+        if action_type.is_registration() {
+            // Complete registration
+            let deposit = PendingRegistrationDeposits::<T>::take(&account_id)
+                .unwrap_or_else(|| parachain_staking::Pallet::<T>::min_collator_stake());
+            
+            let candidate_count = parachain_staking::Pallet::<T>::candidate_pool().0.len() as u32;
+            
+            match parachain_staking::Pallet::<T>::join_candidates(
+                <T as frame_system::Config>::RuntimeOrigin::from(RawOrigin::Signed(
+                    account_id.clone(),
+                )),
+                deposit,
+                candidate_count,
+            ) {
+                Ok(_) => {},
+                Err(e) => {
+                    log::error!(
+                        "Failed to join candidates for {:?}: {:?}",
+                        account_id,
+                        e
+                    );
+                    // Cleanup on failure
+                    if let Some(eth_key) = Self::get_ethereum_public_key_if_exists(&account_id) {
+                        <EthereumPublicKeys<T>>::remove(eth_key);
+                    }
+                    <ValidatorActions<T>>::remove(&account_id, ingress_counter);
+                    return Err(e.error)
+                },
+            }
+
+            // Add to active validators list
+            match <ValidatorAccountIds<T>>::try_append(account_id.clone()) {
```

Comment:
we could also move the happy path of activation (from here until line 776) into a separate function.